hal/src/nRF52840/i2c_hal.cpp

```diff
@@ -235,31 +239,61 @@ static int twi_init(HAL_I2C_Interface i2c) {
     return 0;
 }
 
-void HAL_I2C_Init(HAL_I2C_Interface i2c, void* reserved) {
+static bool HAL_I2C_Config_Is_Valid(const HAL_I2C_Config* config) {
+    if ((config == nullptr) ||
+            (config->rx_buffer == nullptr ||
+             config->rx_buffer_size == 0 ||
+             config->tx_buffer == nullptr ||
+             config->tx_buffer_size == 0)) {
+        return false;
+    }
+
+    return true;
+}
+
+int HAL_I2C_Init(HAL_I2C_Interface i2c, const HAL_I2C_Config* config) {
+    // Disable threading to create the I2C mutex
     os_thread_scheduling(false, NULL);
     if (m_i2c_map[i2c].mutex == NULL) {
         os_mutex_recursive_create(&m_i2c_map[i2c].mutex);
     } else {
         // Already initialized
         os_thread_scheduling(true, NULL);
-        return;
+        return SYSTEM_ERROR_NONE;
     }
 
+    // Capture the mutex and re-enable threading
     HAL_I2C_Acquire(i2c, NULL);
     os_thread_scheduling(true, NULL);
 
+    // Initialize internal data structure
+    if (HAL_I2C_Config_Is_Valid(config)) {
+        m_i2c_map[i2c].rx_buf = config->rx_buffer;
+        m_i2c_map[i2c].rx_buf_size = config->rx_buffer_size;
+        m_i2c_map[i2c].tx_buf = config->tx_buffer;
+        m_i2c_map[i2c].tx_buf_size = config->tx_buffer_size;
+    } else {
+        // Allocate default buffers
+        m_i2c_map[i2c].rx_buf = new (std::nothrow) uint8_t[I2C_BUFFER_LENGTH];
```

Comment:
It might be `nullptr`. Same to `m_i2c_map[i2c].tx_buf`.

Reply:
I agree, though the only sane thing we can probably do is panic in this case.

Reply:
Added assertion.